src/proxy/BaseProxy.ts

```diff
@@ -1,24 +1,49 @@
 import {SerializationService} from '../serialization/SerializationService';
 import {Data} from '../serialization/Data';
+import HazelcastClient = require('../HazelcastClient');
+import ClientMessage = require('../ClientMessage');
+import Q = require('q');
 export class BaseProxy {
 
-    protected client: any;
-    private name: string;
-    private serviceName: string;
+    protected client: HazelcastClient;
+    protected name: string;
+    protected serviceName: string;
 
-    //TODO client should be of type {real client type}
-    constructor(client: { serializationService : SerializationService}, serviceName: string, name: string) {
+    constructor(client: HazelcastClient, serviceName: string, name: string) {
         this.client = client;
         this.name = name;
         this.serviceName = serviceName;
     }
 
+    private createPromise<T>(codec: any, promise: Q.Promise<ClientMessage>): Q.Promise<T> {
+        var deferred: Q.Deferred<T> = Q.defer<T>();
+        var toObject = this.toObject.bind(this);
+        promise.then(function(clientMessage: ClientMessage) {
+            var parameters: any = codec.decodeResponse(clientMessage, toObject);
+            deferred.resolve(parameters.response);
+        });
+        return deferred.promise;
+    }
+
+    protected encodeInvokeOnKey<T>(codec: any, key: any, ...codecArguments: any[]): Q.Promise<T> {
```

Comment:
Sorry, I indeed replied to wrong comment. Basically, you can inline the invokeOnPartition method